File: 완전탐색(백트래킹, 상태트리와 CUT EDGE), DFS 기초/합이 같은 부분집합.py
"""
import sys
# sys.stdin = open("input.txt")

result = "NO"

def dfs(node):
    global result
    if node == n:
        sum = 0
        for i in range(n):
            if check[i] == 1:
                sum += arr[i]
            else:
                sum -= arr[i]
        if sum == 0:
            result = "YES"
    else:
        check[node] = 1
        dfs(node + 1)
        check[node] = 0
        dfs(node + 1)

if __name__ == "__main__":
    n = int(input())
    arr = list(map(int, input().split()))
    check = [0] * n
    dfs(0)
    print(result)
"""
import sys
# sys.stdin = open("input.txt", 'r')

def dfs(level, sum):
    global switch
    if switch == 1: return 
    if sum > total // 2: return

    if level == n:
        if sum == (total - sum):
            print("YES")
            switch = 1
    else:
        dfs(level + 1, sum + arr[level])
        dfs(level + 1, sum)

if __name__ == "__main__":
    switch = 0
    n = int(input())
    arr = list(map(int, input().split()))
    total = sum(arr)
    dfs(0, 0)
    if switch == 0: print("NO")

File: 재귀 & 깊이,넓이 우선탐색/합이 같은 부분집합.cpp
#include <stdio.h>
#include <vector>
using namespace std;

bool flag = false;
int n, total;
vector<int> arr;

void dfs(int level, int sum){
	if(flag == true) return;
	
	if(level == n){
		if(sum == total - sum){
			flag = true;
			return;
		}
	}
	else{
		dfs(level + 1, sum + arr[level]);
		dfs(level + 1, sum);	
	}
}

int main(void){
	// freopen("input.txt", "r", stdin);
	int i, tmp;
	scanf("%d", &n);
	
	for(i = 0; i < n; i++){
		scanf("%d", &tmp);
		arr.push_back(tmp);
		total += arr[i];
	}
	
	dfs(0, 0);
	
	if(flag == true) printf("YES");
	else			 printf("NO");
	
	return 0;
}
